File: ekosis/state_keepers/statistics_keeper.py
import asyncio
import copy
import logging
import numpy as np
import time

from typing import Any, Dict, List, Tuple

from ..util import SingletonType
from ..queues import PaginatedQueue
from ..configuration.config_models import AppConfiguration

# --------------------------------------------------------------------------------
class StatisticsKeeper(metaclass=SingletonType):
    __config              : AppConfiguration          = AppConfiguration()
    __running             : bool                      = False
    __logger              : logging.Logger            = logging.getLogger()
    __gather_period       : int                       = 300
    __history_length      : int                       = 12
    __start_time          : float                     = time.time()
    __statistics_current  : Dict[str, Any]            = {}
    __statistics_history  : List[Dict[str, Any]]      = []
    __persisted_queues    : Dict[str, PaginatedQueue] = {}
    __endpoint_durations  : Dict[str, List[float]]    = {}

    def __init__(self):
        pass

    def set_gather_period(self, gather_period: int):
        self.__gather_period = gather_period

    def set_history_length(self, history_length: int):
        self.__history_length = history_length

    async def get_endpoint_percentiles(self) -> Dict[str, Tuple[float, float]]:
        retval: Dict[str, Tuple[float, float]] = {}
        for key, duration_list in self.__endpoint_durations.items():
            if len(duration_list) > 0:
                p95 = float(np.percentile(duration_list, 95))
                p99 = float(np.percentile(duration_list, 99))
                retval[key] = (p95, p99)
            else:
                retval[key] = (-1.0, -1.0)
        return retval

    def __reset_endpoint_durations(self):
        for key in self.__endpoint_durations.keys():
            self.__endpoint_durations[key] = []

    async def __update_current_statistics(self):
        current_time = time.time()

        self.__statistics_current['timestamp']               = current_time
        self.__statistics_current['uptime']                  = current_time - self.__start_time
        self.__statistics_current['application']             = {}
        self.__statistics_current['application']['name']     = self.__config.name
        self.__statistics_current['application']['instance'] = self.__config.instance

        endpoint_percentiles = await self.get_endpoint_percentiles()
        for key, percentiles in endpoint_percentiles.items():
            self.set_statistic_value(f"endpoint_data.{key}.p95", float(percentiles[0]))
            self.set_statistic_value(f"endpoint_data.{key}.p99", float(percentiles[1]))

        for key in self.__persisted_queues.keys():
            self.set_statistic_value(key, self.__persisted_queues[key].size())

    async def get_current_statistics(self) -> Dict[str, Any]:
        await self.__update_current_statistics()
        return self.__statistics_current

    async def get_last_gathered_statistics(self) -> Dict[str, Any]:
        if len(self.__statistics_history) > 0:
            return self.__statistics_history[0]
        else:
            return {}

    async def get_full_gathered_statistics(self) -> List[Dict[str, Any]]:
        return self.__statistics_history

    def set_statistic_value(self, key: str, value: float):
        keys = key.split('.')
        self.__deep_set(self.__statistics_current, keys, value)

    @staticmethod
    def __deep_get(dictionary, keys):
        for key in keys:
            if isinstance(dictionary, dict):
                dictionary = dictionary.get(key, None)
            else:
                return None
        return dictionary

    @staticmethod
    def __deep_set(dictionary, keys, value):
        for key in keys[:-1]:
            dictionary = dictionary.setdefault(key, {})
        dictionary[keys[-1]] = value

    def __get_statistic_value(self, dictionary, keys) -> float:
        value = self.__deep_get(dictionary, keys)
        if value is None:
            return 0
        else:
            return value

    def increment(self, key: str, value: float = 1):
        keys          = key.split('.')
        current_value = self.__get_statistic_value(self.__statistics_current, keys)
        self.__deep_set(self.__statistics_current, keys, current_value+value)

    def decrement(self, key: str, value: float = 1):
        keys          = key.split('.')
        current_value = self.__get_statistic_value(self.__statistics_current, keys)
        self.__deep_set(self.__statistics_current, keys, current_value-value)

    def add_persisted_queue(self, key: str, queue: PaginatedQueue):
        self.__persisted_queues[key] = queue

    def track_endpoint_data(self, key: str):
        self.increment(f"endpoint_data.{key}.call_count", 0)
        self.__endpoint_durations[key] = []

    def add_endpoint_stats(self, key: str, percentile: float = 0):
        self.increment(f"endpoint_data.{key}.call_count")
        self.__endpoint_durations[key].append(percentile)

    async def __aenter__(self):
        self.__start_time = int(time.time())
        self.__running    = True
        self.__logger.info("Starting stats gathering.")

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        self.__running = False
        self.__logger.info("Stopping stats gathering.")

    def __reset_stats(self, dictionary):
        for key, value in dictionary.items():
            if isinstance(value, dict):
                self.__reset_stats(value)
            else:
                dictionary[key] = 0

    async def __reset_current_statistics(self):
        self.__reset_endpoint_durations()
        self.__reset_stats(self.__statistics_current)

    async def gather_statistics(self):
        while self.__running:
            sleep_count: int = 0

            while self.__running and sleep_count < self.__gather_period:
                await asyncio.sleep(1)
                sleep_count += 1

            if self.__running:
                self.__logger.info(f"Gathering statistics")
                await self.__update_current_statistics()
                self.__statistics_history.insert(0, copy.deepcopy(self.__statistics_current))
                if len(self.__statistics_history) > self.__history_length:
                    self.__statistics_history.pop()
                await self.__reset_current_statistics()

File: ekocpp/src/state_keepers/statistics_keeper.cpp
#include "statistics_keeper.hpp"

#include <algorithm>
#include <chrono>
#include <cmath>

#include <spdlog/spdlog.h>

#include "../configuration/config_models.hpp"

namespace {

// --------------------------------------------------------------------------------
// "endpoint_data.echo.p95" -> "/endpoint_data/echo/p95" -- nlohmann::json's
// json_pointer does the deep get/set itself once converted, replacing
// Python's hand-rolled recursive __deep_get/__deep_set walk.
nlohmann::json::json_pointer to_pointer(const std::string& dotted_key) {
    std::string path = dotted_key;
    std::replace(path.begin(), path.end(), '.', '/');
    return nlohmann::json::json_pointer("/" + path);
}

// --------------------------------------------------------------------------------
double now_seconds() {
    return std::chrono::duration<double>(std::chrono::system_clock::now().time_since_epoch()).count();
}

} // namespace

// --------------------------------------------------------------------------------
StatisticsKeeper& StatisticsKeeper::instance() {
    static StatisticsKeeper keeper;
    return keeper;
}

// --------------------------------------------------------------------------------
void StatisticsKeeper::increment(const std::string& key, double value) {
    const auto pointer  = to_pointer(key);
    const double current = statistics_current_.contains(pointer) ? statistics_current_.at(pointer).get<double>() : 0.0;
    statistics_current_[pointer] = current + value;
}

// --------------------------------------------------------------------------------
void StatisticsKeeper::decrement(const std::string& key, double value) {
    increment(key, -value);
}

// --------------------------------------------------------------------------------
void StatisticsKeeper::set_statistic_value(const std::string& key, double value) {
    statistics_current_[to_pointer(key)] = value;
}

// --------------------------------------------------------------------------------
void StatisticsKeeper::add_persisted_queue(const std::string& key, std::function<size_t()> size_getter) {
    persisted_queues_[key] = std::move(size_getter);
}

// --------------------------------------------------------------------------------
void StatisticsKeeper::track_endpoint_data(const std::string& key) {
    increment("endpoint_data." + key + ".call_count", 0.0);
    endpoint_durations_[key] = {};
}

// --------------------------------------------------------------------------------
void StatisticsKeeper::add_endpoint_stats(const std::string& key, double duration_seconds) {
    increment("endpoint_data." + key + ".call_count");
    endpoint_durations_[key].push_back(duration_seconds);
}

// --------------------------------------------------------------------------------
// Nearest-rank percentile, not Python's statistics.quantiles(n=100) exclusive
// method -- this is a deliberate, known divergence (see statistics_keeper.hpp),
// not an attempt at bit-for-bit parity. Fine for an internal dashboard stat.
StatisticsKeeper::EndpointPercentiles StatisticsKeeper::percentiles_for(std::vector<double> durations) {
    if (durations.empty()) return {-1.0, -1.0};
    if (durations.size() == 1) return {durations[0], durations[0]};

    std::sort(durations.begin(), durations.end());
    auto rank = [&durations](double percentile) {
        const size_t index = static_cast<size_t>(std::ceil(percentile / 100.0 * static_cast<double>(durations.size()))) - 1;
        return durations[std::min(index, durations.size() - 1)];
    };
    return {rank(95.0), rank(99.0)};
}

// --------------------------------------------------------------------------------
void StatisticsKeeper::update_current_statistics() {
    const double current_time = now_seconds();

    statistics_current_["timestamp"]               = current_time;
    statistics_current_["uptime"]                  = current_time - start_time_;
    statistics_current_["gather_period"]           = gather_period_;
    statistics_current_["application"]["name"]     = AppConfiguration::instance().name();
    statistics_current_["application"]["instance"] = AppConfiguration::instance().instance_id();

    for (const auto& [key, durations] : endpoint_durations_) {
        const EndpointPercentiles percentiles = percentiles_for(durations);
        set_statistic_value("endpoint_data." + key + ".p95", percentiles.p95);
        set_statistic_value("endpoint_data." + key + ".p99", percentiles.p99);
    }

    for (const auto& [key, size_getter] : persisted_queues_) {
        set_statistic_value(key, static_cast<double>(size_getter()));
    }
}

// --------------------------------------------------------------------------------
nlohmann::json StatisticsKeeper::get_current_statistics() {
    update_current_statistics();
    return statistics_current_;
}

// --------------------------------------------------------------------------------
const nlohmann::json& StatisticsKeeper::get_last_gathered_statistics() const {
    static const nlohmann::json empty = nlohmann::json::object();
    return statistics_history_.empty() ? empty : statistics_history_.front();
}

// --------------------------------------------------------------------------------
const std::vector<nlohmann::json>& StatisticsKeeper::get_full_gathered_statistics() const {
    return statistics_history_;
}

// --------------------------------------------------------------------------------
// Zeroes every leaf value regardless of type -- matches Python's
// __reset_stats exactly, including the quirk that it also zeroes
// non-numeric leaves like application.name/instance until the next
// update_current_statistics() call repopulates them. Harmless in practice:
// nothing reads statistics_current_ between a reset and the next update.
void StatisticsKeeper::reset_stats(nlohmann::json& node) {
    for (auto& [key, value] : node.items()) {
        if (value.is_object()) {
            reset_stats(value);
        } else {
            value = 0;
        }
    }
}

// --------------------------------------------------------------------------------
void StatisticsKeeper::gather_now() {
    spdlog::info("Gathering statistics");
    update_current_statistics();

    statistics_history_.insert(statistics_history_.begin(), statistics_current_);
    if (statistics_history_.size() > history_length_) {
        statistics_history_.pop_back();
    }

    for (auto& [key, durations] : endpoint_durations_) {
        durations.clear();
    }
    reset_stats(statistics_current_);
}

// --------------------------------------------------------------------------------
void StatisticsKeeper::start() {
    start_time_ = now_seconds();
    running_    = true;
    spdlog::info("Starting stats gathering.");
}

// --------------------------------------------------------------------------------
void StatisticsKeeper::stop() {
    running_ = false;
    spdlog::info("Stopping stats gathering.");
}
